runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapperTest.java

```diff
@@ -690,6 +701,68 @@ private static void testSourceDoesNotShutdown(boolean shouldHaveReaders) throws
       }
       assertThat(thread.isAlive(), is(false));
     }
+
+    @Test
+    public void testSequentialReadingFromBoundedSource() throws Exception {
+      UnboundedReadFromBoundedSource.BoundedToUnboundedSourceAdapter source =
+          new UnboundedReadFromBoundedSource.BoundedToUnboundedSourceAdapter<>(
+              new BoundedIntegerSource(0, 1000));
+
+      FlinkPipelineOptions options = PipelineOptionsFactory.as(FlinkPipelineOptions.class);
+      options.setShutdownSourcesOnFinalWatermark(true);
+
+      UnboundedSourceWrapper<Integer, TestCountingSource.CounterMark> sourceWrapper =
+          new UnboundedSourceWrapper<>("sequentialRead", options, source, 4);
+      StreamingRuntimeContext runtimeContextMock = Mockito.mock(StreamingRuntimeContext.class);
+
+      Mockito.when(runtimeContextMock.getIndexOfThisSubtask()).thenReturn(0);
+      when(runtimeContextMock.getNumberOfParallelSubtasks()).thenReturn(2);
+      when(runtimeContextMock.getExecutionConfig()).thenReturn(new ExecutionConfig());
+
+      TestProcessingTimeService processingTimeService = new TestProcessingTimeService();
+      processingTimeService.setCurrentTime(0);
+      when(runtimeContextMock.getProcessingTimeService()).thenReturn(processingTimeService);
+
+      when(runtimeContextMock.getMetricGroup()).thenReturn(new UnregisteredMetricsGroup());
+
+      sourceWrapper.setRuntimeContext(runtimeContextMock);
+
+      sourceWrapper.open(new Configuration());
+      assertThat(sourceWrapper.getLocalReaders().size(), is(2));
+
+      List<Integer> integers = new ArrayList<>();
+      sourceWrapper.run(
+          new SourceFunction.SourceContext<WindowedValue<ValueWithRecordId<Integer>>>() {
+
+            @Override
+            public void collect(WindowedValue<ValueWithRecordId<Integer>> element) {
+              integers.add(element.getValue().getValue());
+            }
+
+            @Override
+            public void collectWithTimestamp(
+                WindowedValue<ValueWithRecordId<Integer>> element, long timestamp) {
+              throw new IllegalStateException("Should not collect with timestamp");
+            }
+
+            @Override
+            public void emitWatermark(Watermark mark) {}
+
+            @Override
+            public void markAsTemporarilyIdle() {}
+
+            @Override
+            public Object getCheckpointLock() {
+              return new Object();
+            }
+
+            @Override
+            public void close() {}
+          });
+
+      // the source is effectively split into two parts
+      assertThat(integers.size(), is(500));
```

Comment:
If you look further up, it sets the number of parallel sub tasks to 2 and the initial splitting is performed with a parallelism of 4. So that makes 2 instances for this partition which is run in the test.

Good point on the output ordering. I've improved the output check. Please have a look.

Reply:
Aha, I missed `Mockito.when(runtimeContextMock.getIndexOfThisSubtask()).thenReturn(0)` .. I thought it is going to process both partitions. Thanks